crypto/rand/drbg_hash.c

```diff
@@ -0,0 +1,292 @@
+                && EVP_DigestUpdate(ctx, tmp, tmp_sz)
+                && EVP_DigestUpdate(ctx, in, inlen)
+                && (in2 == NULL || EVP_DigestUpdate(ctx, in2, in2len))
+                && (in3 == NULL || EVP_DigestUpdate(ctx, in3, in3len))))
+            return 0;
+
+        if (outlen < hash->blocklen) {
+            if (!EVP_DigestFinal(ctx, vtmp, NULL))
+                return 0;
+            memcpy(out, vtmp, outlen);
+            OPENSSL_cleanse(vtmp, hash->blocklen);
+            break;
+        }
+        else if(!EVP_DigestFinal(ctx, out, NULL))
+            return 0;
+
+        outlen -= hash->blocklen;
+        if (outlen == 0)
+            break;
+        /* (Step 4.2) counter++ */
+        tmp[0]++;
+        out += hash->blocklen;
+    }
+    return 1;
+}
+
+static int hash_df1(RAND_DRBG *drbg, unsigned char *out,
+                    const unsigned char in_byte,
+                    const unsigned char *in1, size_t in1len)
+{
+    return hash_df(drbg, out, in_byte, in1, in1len, NULL, 0, NULL, 0);
+}
+
+/*
+ * Add 2 byte buffers together. The first elements in each buffer are the top
+ * most bytes. The result is stored in the dst buffer.
+ * The final carry is ignored i.e: dst =  (dst + in) mod (2^seedlen_bits).
+ * where dst size is drbg->seedlen, and inlen <= drbg->seedlen.
+ */
+static int add_bytes(RAND_DRBG *drbg, unsigned char *dst,
+                     unsigned char *in, size_t inlen)
+{
+    size_t i;
+    int result;
+    const unsigned char *add;
+    unsigned char carry = 0, *d;
+
+    if (drbg->seedlen < 1 || inlen < 1 || inlen > drbg->seedlen)
+        return 0;
+
+    d = &dst[drbg->seedlen - 1];
+    add = &in[inlen - 1];
+
+    for (i = inlen; i > 0; i--, d--, add--) {
+        result = *d + *add + carry;
+        carry = (unsigned char)(result >> 8);
+        *d = (result & 0xff);
+    }
+
+    if (carry != 0) {
+        /* Add the carry to the top of the dst if inlen is not the same size */
+        for (i = drbg->seedlen - inlen; i > 0; --i, d--) {
+            *d += 1;     /* Carry can only be 1 */
+            if (*d != 0) /* exit if carry doesnt propagate to the next byte */
+                break;
+        }
+    }
+    return 1;
+}
+
+/* V = (V + Hash(inbyte || V { || additional_input}) mod (2^seedlen) */
+static int add_hash_to_v(RAND_DRBG *drbg, unsigned char inbyte,
+                         const unsigned char *adin, size_t adinlen)
+{
+    RAND_DRBG_HASH *hash = &drbg->data.hash;
+    EVP_MD_CTX *ctx = hash->ctx;
+    return EVP_DigestInit_ex(ctx, hash->md, NULL)
+           && EVP_DigestUpdate(ctx, &inbyte, 1)
+           && EVP_DigestUpdate(ctx, hash->V, drbg->seedlen)
+           && (adin == NULL || EVP_DigestUpdate(ctx, adin, adinlen))
+           && EVP_DigestFinal(hash->ctx, hash->vtmp, NULL)
+           && add_bytes(drbg, hash->V, hash->vtmp, hash->blocklen);
+}
+
+static int hash_gen(RAND_DRBG *drbg, unsigned char *out, size_t outlen)
```

Comment:
A comment describing the function would be helpful (and maybe a few more in the function itself)